crates/biome_js_analyze/src/lint/nursery/no_yoda_expression.rs

```diff
@@ -0,0 +1,479 @@
+) -> bool {
+    let left_identifier = left_binary_expression.right().ok();
+    let right_identifier = right_binary_expression.left().ok();
+
+    if !matches!(operator, JsLogicalOperator::LogicalAnd)
+        || !is_same_identifier(&left_identifier, &right_identifier)
+    {
+        return false;
+    }
+
+    let left_literal = extract_string_value(left_binary_expression.left().as_ref().ok());
+    let right_literal = extract_string_value(right_binary_expression.right().as_ref().ok());
+
+    match (left_literal, right_literal) {
+        (None, None) => false,
+        (Some(_), None) => true,
+        (None, Some(_)) => true,
+        (Some(left_value), Some(right_value)) => {
+            compare_literals(left_value.as_str(), right_value.as_str())
+        }
+    }
+}
+
+/// Compare literals as number if possible, compare as string otherwise
+fn compare_literals(left_string_value: &str, right_string_value: &str) -> bool {
+    match (
+        left_string_value.parse::<f64>(),
+        right_string_value.parse::<f64>(),
+    ) {
+        (Ok(left_number_value), Ok(right_number_value)) => left_number_value <= right_number_value,
+        _ => left_string_value <= right_string_value,
+    }
+}
+
+/// Determines whether the nodes compose an outside range assertion, like `x < 0 || 1 <= x`
+fn is_outside_range_assertion(
+    operator: &JsLogicalOperator,
+    left_binary_expression: &JsBinaryExpression,
+    right_binary_expression: &JsBinaryExpression,
+) -> bool {
+    let left_identifier = left_binary_expression.left().ok();
+    let right_identifier = right_binary_expression.right().ok();
+
+    if !matches!(operator, JsLogicalOperator::LogicalOr)
+        || !is_same_identifier(&left_identifier, &right_identifier)
+    {
+        return false;
+    }
+
+    let left_literal = extract_string_value(left_binary_expression.right().as_ref().ok());
+    let right_literal = extract_string_value(right_binary_expression.left().as_ref().ok());
+
+    match (left_literal, right_literal) {
+        (None, None) => false,
+        (Some(_), None) => true,
+        (None, Some(_)) => true,
+        (Some(left_value), Some(right_value)) => {
+            compare_literals(left_value.as_str(), right_value.as_str())
+        }
+    }
+}
+
+fn is_wrapped_in_parenthesis(logical_expression: &JsLogicalExpression) -> bool {
+    !matches!(
+        (
+            logical_expression.parent::<JsParenthesizedExpression>(),
+            logical_expression.parent::<JsIfStatement>(),
+            logical_expression.parent::<JsWhileStatement>(),
+        ),
+        (None, None, None)
+    )
+}
+
+fn is_same_identifier(
+    left_expression: &Option<AnyJsExpression>,
+    right_expression: &Option<AnyJsExpression>,
+) -> bool {
+    match (left_expression, right_expression) {
+        (Some(left), Some(right)) => {
+            // We can't consider two call expressions equal here because the result of the expression might be different if we call it twice
```

Comment:
I've updated the documentation of this method to explain it better why it's needed, but I had to do this check to ensure that this method will return false when there's a call expression inside the expression. 

Consider this expression: `if (0 <= a[b()] && a[b()] < 1) {}`

We can't assume the expressions are equal because the result of the `JsCallExpression` might be different in multiple calls:
![image](https://github.com/biomejs/biome/assets/11337615/181b9141-bc21-4997-96ea-93e2878e2909)

Reply:
I see. I think it is a small thing. It is most likely to cause false positives. I assume you ported it from the ESLint rule? Instead, I could remove this check and just recognize this as a range. What do you think @ematipico ?

Reply:
Expressions can't be statically evaluated (unless they are literal), so we should skip them.

We do the same in the rule `noSelfAssignment`: even though left and right call the same expressions, we can't know their result.

Reply:
What do you mean by `skip them`?

Reply:
I mean we shouldn't trigger the rule

Reply:
I see your point. However, I think it is a bit different: being conservative for `noSelfAssignment` allows us to avoid false positives. Conversely, being conservative on range detection led us to false positives. In this case, I think being permissive on range detection avoids false positives.

Reply:
In this case `if (0 <= a[b()] && a[b()] < 1) {}` the rule won't recognize the expression as a range, but `0 <= a[b()]` alone still is a yoda expression, so it will trigger the rule anyways, are you ok with that @ematipico?

Reply:
I think that @ematipico means that every expression should be considered as different. For example `0 < o.p && o.p < 10` would not be recognized as a range because `o.p` can give different results.

Conversely, I think we could recognize all cases, including `(0 <= a[b()] && a[b()] < 1)` as ranges.

Reply:
@michellocana let's accept `(0 <= a[b()] && a[b()] < 1)` and I think we can merge.

Reply:
done! @Conaclos